Sources/SwiftOperatorPrecedence/OperatorPrecedence+Folding.swift

```diff
@@ -0,0 +1,364 @@
+  }
+
+  /// Determine whether we should consider an operator in the given group
+  /// based on the specified bound.
+  private func shouldConsiderOperator(
+    fromGroup groupName: PrecedenceGroupName?,
+    in bound: PrecedenceBound,
+    fromGroupSyntax: SyntaxProtocol?,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows -> Bool {
+    guard let boundGroupName = bound.groupName else {
+      return true
+    }
+
+    guard let groupName = groupName else {
+      return false
+    }
+
+    if groupName == boundGroupName {
+      return !bound.isStrict
+    }
+
+    return try precedence(
+      relating: groupName, to: boundGroupName,
+      startSyntax: fromGroupSyntax, endSyntax: bound.syntax,
+      errorHandler: errorHandler
+    ) != .lowerThan
+  }
+
+  /// Look up the precedence group for the given expression syntax.
+  private func lookupPrecedence(
+    of expr: ExprSyntax,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows -> PrecedenceGroupName? {
+    // A binary operator.
+    if let binaryExpr = expr.as(BinaryOperatorExprSyntax.self) {
+      let operatorName = binaryExpr.operatorToken.text
+      return try lookupOperatorPrecedenceGroupName(
+        operatorName, referencedFrom: binaryExpr.operatorToken,
+        errorHandler: errorHandler
+      )
+    }
+
+    // FIXME: Handle all of the language-defined precedence relationships.
+    return nil
+  }
+
+  /// Form a binary operation expression for, e.g., a + b.
+  private func makeBinaryOperationExpr(
+    lhs: ExprSyntax, op: ExprSyntax, rhs: ExprSyntax
+  ) -> ExprSyntax {
+    ExprSyntax(
+      InfixOperatorExprSyntax(
+        leftOperand: lhs, operatorOperand: op, rightOperand: rhs)
+      )
+  }
+
+  /// Determine the associativity between two precedence groups.
+  private func associativity(
+    firstGroup: PrecedenceGroupName?,
+    firstGroupSyntax: SyntaxProtocol?,
+    secondGroup: PrecedenceGroupName?,
+    secondGroupSyntax: SyntaxProtocol?,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows -> Associativity {
+    guard let firstGroup = firstGroup, let secondGroup = secondGroup else {
+      return .none
+    }
+
+    // If we have the same group, query its associativity.
+    if firstGroup == secondGroup {
+      guard let group = precedenceGraph.lookupGroup(firstGroup) else {
+        try errorHandler(
+          .missingGroup(firstGroup, referencedFrom: firstGroupSyntax))
+        return .none
+      }
+
+      return group.associativity
+    }
+
+    if try precedence(
+      relating: firstGroup, to: secondGroup,
+      startSyntax: firstGroupSyntax, endSyntax: secondGroupSyntax,
+      errorHandler: errorHandler
+    ) == .higherThan {
+      return .left
+    }
+
+    if try precedence(
+      relating: secondGroup, to: firstGroup,
+      startSyntax: secondGroupSyntax, endSyntax: firstGroupSyntax,
+      errorHandler: errorHandler
+    ) == .higherThan {
```

Comment:
Isn’t `precedence` guaranteed to be commutative. I.e. should’t this check be the same as checking that the `precedence` call in the `if`-statement above returned `.lowerThan`?